src/rpc/core_rpc_server.cpp

```diff
@@ -928,6 +1042,12 @@ namespace cryptonote
   bool core_rpc_server::on_getblockhash(const COMMAND_RPC_GETBLOCKHASH::request& req, COMMAND_RPC_GETBLOCKHASH::response& res, epee::json_rpc::error& error_resp)
   {
     PERF_TIMER(on_getblockhash);
+    boost::lock_guard<boost::mutex> lock(m_bootstrap_daemon_mutex);
+    if (m_should_use_bootstrap_daemon)
+    {
+      res = "This command is unsupported for bootstrap daemon";
+      return false;
+    }
```

Comment:
That does lock the whole RPC though.

Reply:
That's why there should be a dedicated mutex, not the one responsible for locking RPC requests.

Reply:
I’m not sure what you mean by “a dedicated mutex”. The locking here is supposed to lock the whole RPC when `m_bootstrap_daemon_address` isn’t empty. This is the cost we have to pay in any case, no?

Reply:
Ah, I did not realize you intended this, all good then. Is this to avoid out of order responses ? That's a good point since you could then end up with an untrusted RPC returned after a switch to trusted...

Reply:
The problem I'm seeing here is that currently you acquire an exclusive lock on `m_bootstrap_daemon_mutex` for the whole call to e.g. `core_rpc_server::on_getblockhash`. This degrades performance quite a bit as you keep the lock for longer than required. This could be solved by either using two different mutexes (one for `m_was_bootstrap_ever_used`, another for `m_should_use_bootstrap_daemon`) although this might end up being a bit too complicated.

It already helps if you just try to keep the mutex for as short as possible:

```cpp
m_bootstrap_daemon_mutex.lock();
res.was_bootstrap_ever_used = m_was_bootstrap_ever_used;
m_bootstrap_daemon_mutex.unlock();
```

and

```cpp
bool use_bootstrap_daemon;
m_bootstrap_daemon_mutex.lock();
use_bootstrap_daemon = m_should_use_bootstrap_daemon;
m_bootstrap_daemon_mutex.unlock();

if (use_bootstrap_daemon)
{
      res = "This command is unsupported for bootstrap daemon";
      return false;
}
```

Even better than using a `boost::mutex` (which only supports `locking`) you could use a `boost::shared_mutex` which supports read & write locking (a semaphore). That is, when being read-locked (`shared_lock`), another thread trying to read-lock it doesn't have to wait for it to be unlocked again (as multiple readers are allowed). However, when being write-locked (`unique_lock`), no other reader / writer can acquire the lock. This improves performance as it allows for multiple concurrent readers.

__Note__: I haven't done any benchmarks of the current implementation, so I don't know how much of a performance issue this currently is.

Reply:
@moneromooo-monero 

Yes, this is to make sure that the state of `m_should_use_bootstrap_daemon` is consistent with other possible threads that are trying to write to it concurrently.

@leonklingele 

I don't see a difference between your suggested code
```c++
m_bootstrap_daemon_mutex.lock();
res.was_bootstrap_ever_used = m_was_bootstrap_ever_used;
m_bootstrap_daemon_mutex.unlock();
```
and my current code
```c++
    {
      boost::lock_guard<boost::mutex> lock(m_bootstrap_daemon_mutex);
      res.was_bootstrap_ever_used = m_was_bootstrap_ever_used;
    }
```

I like your suggestion to use `boost::shared_mutex`, thanks.